File: lidar_processor/lidar_processor/pointcloud_aggregator_node.py
# Copyright (c) 2024, RoboVerse community
# SPDX-License-Identifier: BSD-3-Clause

"""
PointCloud Aggregator Node

Advanced point cloud processing with filtering, downsampling, and visualization.
"""

from typing import List
from dataclasses import dataclass
import time

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy
from sensor_msgs.msg import PointCloud2
from sensor_msgs_py import point_cloud2
import numpy as np


@dataclass 
class AggregatorConfig:
    """Configuration for point cloud aggregator"""
    max_range: float = 20.0        # Maximum range from robot center
    min_range: float = 0.1         # Minimum range from robot center
    height_filter_min: float = -2.0  # Minimum height (z-coordinate)
    height_filter_max: float = 3.0   # Maximum height (z-coordinate)
    downsample_rate: int = 10       # Keep every Nth point
    publish_rate: float = 5.0       # Hz


class StatisticalFilter:
    """Statistical outlier removal filter"""
    
    def __init__(self, k_neighbors: int = 20, std_ratio: float = 2.0):
        self.k_neighbors = k_neighbors
        self.std_ratio = std_ratio
    
    def filter_points(self, points: np.ndarray) -> np.ndarray:
        """Remove statistical outliers from point cloud"""
        if len(points) < self.k_neighbors:
            return points
        
        # Calculate distances to k nearest neighbors for each point
        distances = []
        for i, point in enumerate(points):
            # Calculate distances to all other points
            dists = np.linalg.norm(points - point, axis=1)
            # Get k+1 nearest (including self) and take mean of k neighbors
            nearest_dists = np.partition(dists, self.k_neighbors)[:self.k_neighbors+1]
            distances.append(np.mean(nearest_dists[1:]))  # Exclude self (distance=0)
        
        # Filter based on statistical threshold
        distances = np.array(distances)
        mean_dist = np.mean(distances)
        std_dist = np.std(distances)
        threshold = mean_dist + self.std_ratio * std_dist
        
        # Keep points within threshold
        mask = distances <= threshold
        return points[mask]


class PointCloudAggregatorNode(Node):
    """Advanced point cloud aggregation and processing node"""
    
    def __init__(self):
        super().__init__('pointcloud_aggregator')
        
        # Declare parameters
        self._declare_parameters()
        self.config = self._load_configuration()
        
        # Initialize filters
        self.statistical_filter = StatisticalFilter()
        
        # Setup QoS
        self.qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=5
        )
        
        # Storage for aggregated points
        self.aggregated_points: List[np.ndarray] = []
        self.last_publish_time = time.time()
        
        # Setup subscriptions and publishers
        self._setup_subscriptions()
        self._setup_publishers()
        
        # Publish timer
        self.publish_timer = self.create_timer(
            1.0 / self.config.publish_rate,
            self._publish_callback
        )
        
        self.get_logger().info("🔄 PointCloud Aggregator Node initialized")
        self._log_configuration()
    
    def _declare_parameters(self) -> None:
        """Declare node parameters"""
        self.declare_parameter('max_range', 20.0)
        self.declare_parameter('min_range', 0.1)
        self.declare_parameter('height_filter_min', -2.0)
        self.declare_parameter('height_filter_max', 3.0)
        self.declare_parameter('downsample_rate', 10)
        self.declare_parameter('publish_rate', 5.0)
    
    def _load_configuration(self) -> AggregatorConfig:
        """Load configuration from parameters"""
        return AggregatorConfig(
            max_range=self.get_parameter('max_range').get_parameter_value().double_value,
            min_range=self.get_parameter('min_range').get_parameter_value().double_value,
            height_filter_min=self.get_parameter('height_filter_min').get_parameter_value().double_value,
            height_filter_max=self.get_parameter('height_filter_max').get_parameter_value().double_value,
            downsample_rate=self.get_parameter('downsample_rate').get_parameter_value().integer_value,
            publish_rate=self.get_parameter('publish_rate').get_parameter_value().double_value
        )
    
    def _setup_subscriptions(self) -> None:
        """Setup subscriptions to raw point clouds"""
        self.subscription = self.create_subscription(
            PointCloud2,
            '/pointcloud/aggregated',
            self._pointcloud_callback,
            self.qos_profile
        )
    
    def _setup_publishers(self) -> None:
        """Setup publishers for processed data"""
        self.filtered_pub = self.create_publisher(
            PointCloud2,
            '/pointcloud/filtered',
            self.qos_profile
        )
        
        self.downsampled_pub = self.create_publisher(
            PointCloud2,
            '/pointcloud/downsampled',
            self.qos_profile
        )
    
    def _pointcloud_callback(self, msg: PointCloud2) -> None:
        """Process incoming aggregated point cloud"""
        try:
            # Extract points
            points = []
            for point in point_cloud2.read_points(msg, field_names=("x", "y", "z"), skip_nans=True):
                points.append([point[0], point[1], point[2]])
            
            if not points:
                return
            
            points_array = np.array(points)
            
            # Apply filters
            filtered_points = self._apply_filters(points_array)
            
            # Store for aggregation
            if len(filtered_points) > 0:
                self.aggregated_points.append(filtered_points)
                
                # Keep only recent data (last 10 seconds worth)
                max_clouds = int(self.config.publish_rate * 10)
                if len(self.aggregated_points) > max_clouds:
                    self.aggregated_points = self.aggregated_points[-max_clouds:]
            
        except Exception as e:
            self.get_logger().error(f"Error processing point cloud: {e}")
    
    def _apply_filters(self, points: np.ndarray) -> np.ndarray:
        """Apply all filtering operations"""
        if len(points) == 0:
            return points
        
        # Range filter
        distances = np.linalg.norm(points[:, :2], axis=1)  # XY distance only
        range_mask = (distances >= self.config.min_range) & (distances <= self.config.max_range)
        points = points[range_mask]
        
        if len(points) == 0:
            return points
        
        # Height filter
        height_mask = (points[:, 2] >= self.config.height_filter_min) & \
                     (points[:, 2] <= self.config.height_filter_max)
        points = points[height_mask]
        
        if len(points) == 0:
            return points
        
        # Statistical outlier removal (only if enough points)
        if len(points) > 100:
            points = self.statistical_filter.filter_points(points)
        
        return points
    
    def _publish_callback(self) -> None:
        """Periodic publishing of processed data"""
        try:
            if not self.aggregated_points:
                return
            
            # Combine all stored point clouds
            all_points = np.vstack(self.aggregated_points)
            
            if len(all_points) == 0:
                return
            
            # Create header
            header = self.get_clock().now().to_msg()
            header.frame_id = "base_link"
            
            # Publish filtered cloud
            filtered_msg = point_cloud2.create_cloud_xyz32(header, all_points.tolist())
            self.filtered_pub.publish(filtered_msg)
            
            # Create and publish downsampled version
            if self.config.downsample_rate > 1:
                downsampled_points = all_points[::self.config.downsample_rate]
                downsampled_msg = point_cloud2.create_cloud_xyz32(header, downsampled_points.tolist())
                self.downsampled_pub.publish(downsampled_msg)
            
            # Log statistics periodically
            current_time = time.time()
            if current_time - self.last_publish_time > 5.0:  # Every 5 seconds
                self.get_logger().info(
                    f"📊 Processed: {len(all_points):,} points, "
                    f"Clouds: {len(self.aggregated_points)}, "
                    f"Downsampled: {len(all_points)//max(1, self.config.downsample_rate):,}"
                )
                self.last_publish_time = current_time
            
        except Exception as e:
            self.get_logger().error(f"Error publishing processed data: {e}")
    
    def _log_configuration(self) -> None:
        """Log current configuration"""
        self.get_logger().info("⚙️  Aggregator Configuration:")
        self.get_logger().info(f"   Range filter: {self.config.min_range:.1f} - {self.config.max_range:.1f}m")
        self.get_logger().info(f"   Height filter: {self.config.height_filter_min:.1f} - {self.config.height_filter_max:.1f}m")
        self.get_logger().info(f"   Downsample rate: 1/{self.config.downsample_rate}")
        self.get_logger().info(f"   Publish rate: {self.config.publish_rate:.1f} Hz")


def main(args=None):
    """Main entry point"""
    rclpy.init(args=args)
    
    try:
        node = PointCloudAggregatorNode()
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    except Exception as e:
        print(f"Error running aggregator: {e}")
    finally:
        rclpy.shutdown()


if __name__ == '__main__':
    main() 

File: lidar_processor_cpp/src/pointcloud_aggregator_node.cpp
// Copyright (c) 2024, RoboVerse community
// SPDX-License-Identifier: BSD-3-Clause

#include "lidar_processor_cpp/pointcloud_aggregator_node.hpp"
#include <algorithm>
#include <cmath>

namespace lidar_processor_cpp
{

StatisticalFilter::StatisticalFilter(int k_neighbors, double std_ratio)
  : k_neighbors_(k_neighbors), std_ratio_(std_ratio)
{
  sor_filter_.setMeanK(k_neighbors_);
  sor_filter_.setStddevMulThresh(std_ratio_);
}

pcl::PointCloud<pcl::PointXYZ>::Ptr StatisticalFilter::filterPoints(
  const pcl::PointCloud<pcl::PointXYZ>::Ptr& input_cloud)
{
  if (input_cloud->points.size() < static_cast<size_t>(k_neighbors_)) {
    return input_cloud;
  }
  
  pcl::PointCloud<pcl::PointXYZ>::Ptr filtered_cloud(new pcl::PointCloud<pcl::PointXYZ>);
  
  sor_filter_.setInputCloud(input_cloud);
  sor_filter_.filter(*filtered_cloud);
  
  return filtered_cloud;
}

PointCloudAggregatorNode::PointCloudAggregatorNode()
  : Node("pointcloud_aggregator")
{
  // Declare parameters
  declareParameters();
  config_ = loadConfiguration();
  
  // Initialize filters
  statistical_filter_ = std::make_unique<StatisticalFilter>(20, 2.0);
  
  // Initialize timing
  last_publish_time_ = std::chrono::steady_clock::now();
  
  // Setup subscriptions and publishers
  setupSubscriptions();
  setupPublishers();
  
  // Publish timer
  publish_timer_ = this->create_wall_timer(
    std::chrono::duration<double>(1.0 / config_.publish_rate),
    std::bind(&PointCloudAggregatorNode::publishCallback, this)
  );
  
  RCLCPP_INFO(this->get_logger(), "🔄 PointCloud Aggregator Node initialized");
  logConfiguration();
}

void PointCloudAggregatorNode::declareParameters()
{
  this->declare_parameter("max_range", 20.0);
  this->declare_parameter("min_range", 0.1);
  this->declare_parameter("height_filter_min", -2.0);
  this->declare_parameter("height_filter_max", 3.0);
  this->declare_parameter("downsample_rate", 10);
  this->declare_parameter("publish_rate", 5.0);
}

AggregatorConfig PointCloudAggregatorNode::loadConfiguration()
{
  AggregatorConfig config;
  
  config.max_range = this->get_parameter("max_range").as_double();
  config.min_range = this->get_parameter("min_range").as_double();
  config.height_filter_min = this->get_parameter("height_filter_min").as_double();
  config.height_filter_max = this->get_parameter("height_filter_max").as_double();
  config.downsample_rate = this->get_parameter("downsample_rate").as_int();
  config.publish_rate = this->get_parameter("publish_rate").as_double();
  
  return config;
}

void PointCloudAggregatorNode::setupSubscriptions()
{
  auto qos = rclcpp::QoS(5)
    .reliability(rclcpp::ReliabilityPolicy::BestEffort)
    .history(rclcpp::HistoryPolicy::KeepLast);
  
  subscription_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
    "/pointcloud/aggregated",
    qos,
    std::bind(&PointCloudAggregatorNode::pointcloudCallback, this, std::placeholders::_1)
  );
}

void PointCloudAggregatorNode::setupPublishers()
{
  auto qos = rclcpp::QoS(5)
    .reliability(rclcpp::ReliabilityPolicy::BestEffort)
    .history(rclcpp::HistoryPolicy::KeepLast);
  
  filtered_pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(
    "/pointcloud/filtered", qos
  );
  
  downsampled_pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(
    "/pointcloud/downsampled", qos
  );
}

void PointCloudAggregatorNode::pointcloudCallback(const sensor_msgs::msg::PointCloud2::SharedPtr msg)
{
  try {
    // Convert ROS message to PCL point cloud
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::fromROSMsg(*msg, *cloud);
    
    if (cloud->points.empty()) {
      return;
    }
    
    // Apply filters
    auto filtered_cloud = applyFilters(cloud);
    
    // Store for aggregation
    if (!filtered_cloud->points.empty()) {
      std::lock_guard<std::mutex> lock(clouds_mutex_);
      aggregated_clouds_.push_back(filtered_cloud);
      
      // Keep only recent data (last 10 seconds worth)
      size_t max_clouds = static_cast<size_t>(config_.publish_rate * 10);
      if (aggregated_clouds_.size() > max_clouds) {
        aggregated_clouds_.erase(
          aggregated_clouds_.begin(),
          aggregated_clouds_.begin() + (aggregated_clouds_.size() - max_clouds)
        );
      }
    }
    
  } catch (const std::exception& e) {
    RCLCPP_ERROR(this->get_logger(), "Error processing point cloud: %s", e.what());
  }
}

pcl::PointCloud<pcl::PointXYZ>::Ptr PointCloudAggregatorNode::applyFilters(
  const pcl::PointCloud<pcl::PointXYZ>::Ptr& input_cloud)
{
  if (input_cloud->points.empty()) {
    return input_cloud;
  }
  
  pcl::PointCloud<pcl::PointXYZ>::Ptr filtered_cloud(new pcl::PointCloud<pcl::PointXYZ>);
  
  // Apply range and height filters
  for (const auto& point : input_cloud->points) {
    // Skip invalid points
    if (!std::isfinite(point.x) || !std::isfinite(point.y) || !std::isfinite(point.z)) {
      continue;
    }
    
    // Range filter (XY distance only)
    double distance = std::sqrt(point.x * point.x + point.y * point.y);
    if (distance < config_.min_range || distance > config_.max_range) {
      continue;
    }
    
    // Height filter
    if (point.z < config_.height_filter_min || point.z > config_.height_filter_max) {
      continue;
    }
    
    filtered_cloud->points.push_back(point);
  }
  
  // Update cloud properties
  filtered_cloud->width = filtered_cloud->points.size();
  filtered_cloud->height = 1;
  filtered_cloud->is_dense = true;
  
  // Statistical outlier removal (only if enough points)
  if (filtered_cloud->points.size() > 100) {
    filtered_cloud = statistical_filter_->filterPoints(filtered_cloud);
  }
  
  return filtered_cloud;
}

void PointCloudAggregatorNode::publishCallback()
{
  try {
    std::lock_guard<std::mutex> lock(clouds_mutex_);
    
    if (aggregated_clouds_.empty()) {
      return;
    }
    
    // Combine all stored point clouds
    pcl::PointCloud<pcl::PointXYZ>::Ptr combined_cloud(new pcl::PointCloud<pcl::PointXYZ>);
    
    for (const auto& cloud : aggregated_clouds_) {
      *combined_cloud += *cloud;
    }
    
    if (combined_cloud->points.empty()) {
      return;
    }
    
    // Update cloud properties
    combined_cloud->width = combined_cloud->points.size();
    combined_cloud->height = 1;
    combined_cloud->is_dense = true;
    
    // Create header
    std_msgs::msg::Header header;
    header.stamp = this->get_clock()->now();
    header.frame_id = "base_link";
    
    // Publish filtered cloud
    sensor_msgs::msg::PointCloud2 filtered_msg;
    pcl::toROSMsg(*combined_cloud, filtered_msg);
    filtered_msg.header = header;
    filtered_pub_->publish(filtered_msg);
    
    // Create and publish downsampled version
    if (config_.downsample_rate > 1) {
      pcl::PointCloud<pcl::PointXYZ>::Ptr downsampled_cloud(new pcl::PointCloud<pcl::PointXYZ>);
      
      for (size_t i = 0; i < combined_cloud->points.size(); i += config_.downsample_rate) {
        downsampled_cloud->points.push_back(combined_cloud->points[i]);
      }
      
      downsampled_cloud->width = downsampled_cloud->points.size();
      downsampled_cloud->height = 1;
      downsampled_cloud->is_dense = true;
      
      sensor_msgs::msg::PointCloud2 downsampled_msg;
      pcl::toROSMsg(*downsampled_cloud, downsampled_msg);
      downsampled_msg.header = header;
      downsampled_pub_->publish(downsampled_msg);
    }
    
    // Log statistics periodically
    auto current_time = std::chrono::steady_clock::now();
    auto time_diff = std::chrono::duration_cast<std::chrono::seconds>(
      current_time - last_publish_time_).count();
    
    if (time_diff >= 5) {  // Every 5 seconds
      size_t total_points = combined_cloud->points.size();
      size_t downsampled_points = total_points / std::max(1, config_.downsample_rate);
      
      RCLCPP_INFO(this->get_logger(),
        "📊 Processed: %zu points, Clouds: %zu, Downsampled: %zu",
        total_points, aggregated_clouds_.size(), downsampled_points);
      
      last_publish_time_ = current_time;
    }
    
  } catch (const std::exception& e) {
    RCLCPP_ERROR(this->get_logger(), "Error publishing processed data: %s", e.what());
  }
}

void PointCloudAggregatorNode::logConfiguration()
{
  RCLCPP_INFO(this->get_logger(), "⚙️  Aggregator Configuration:");
  RCLCPP_INFO(this->get_logger(), "   Range filter: %.1f - %.1fm", 
    config_.min_range, config_.max_range);
  RCLCPP_INFO(this->get_logger(), "   Height filter: %.1f - %.1fm", 
    config_.height_filter_min, config_.height_filter_max);
  RCLCPP_INFO(this->get_logger(), "   Downsample rate: 1/%d", config_.downsample_rate);
  RCLCPP_INFO(this->get_logger(), "   Publish rate: %.1f Hz", config_.publish_rate);
}

}  // namespace lidar_processor_cpp

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  
  try {
    auto node = std::make_shared<lidar_processor_cpp::PointCloudAggregatorNode>();
    rclcpp::spin(node);
  } catch (const std::exception& e) {
    std::cerr << "Error running aggregator: " << e.what() << std::endl;
    return 1;
  }
  
  rclcpp::shutdown();
  return 0;
}